libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1037 @@
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (AppendStoresTaskData,
+                               append_stores_task_data_free);
+
+static void
+append_stores_task_from_proxy_thread (GTask        *task,
+                                      gpointer      source,
+                                      gpointer      task_data,
+                                      GCancellable *cancellable)
+{
+  AppendStoresTaskData *data = task_data;
+  g_autoptr(GError) local_error = NULL;
+
+  gpointer results = data->proxy_func (data->ka_proxy,
+                                       data->proxy_data,
+                                       cancellable,
+                                       &local_error);
+
+  if (results == NULL)
+    {
+      g_task_return_error (task, g_steal_pointer (&local_error));
+      return;
+    }
+
+  g_task_return_pointer (task, g_steal_pointer (&results), data->proxy_return_destroy);
+}
+
+static void
+append_stores_task_from_proxy (EosDiscoveryFeedKnowledgeAppProxy *ka_proxy,
+                               AppendStoresFromProxyFunc          proxy_func,
+                               GDestroyNotify                     proxy_return_destroy,
+                               gpointer                           proxy_func_data,
+                               GCancellable                      *cancellable,
+                               GAsyncReadyCallback                callback,
+                               gpointer                           user_data)
+{
+  g_autoptr(GTask) task = g_task_new (NULL, cancellable, callback, user_data);
+  g_autoptr(AppendStoresTaskData) data = append_stores_task_data_new (ka_proxy,
+                                                                      proxy_func,
+                                                                      proxy_return_destroy,
+                                                                      proxy_func_data);
+
+  g_task_set_return_on_cancel (task, TRUE);
+  g_task_set_task_data (task,
+                        g_steal_pointer (&data),
+                        (GDestroyNotify) append_stores_task_data_free);
+  g_task_run_in_thread (task, append_stores_task_from_proxy_thread);
+}
+
+static gpointer
+ref_object_copy_func (gconstpointer src,
+                      gpointer      user_data)
+{
+  /* Need to cast back to gpointer as our way of "copying" is to increment
+   * the ref-count */
+  return g_object_ref ((gpointer) src);
+}
+
+static GPtrArray *
+flat_map_ptr_arrays (GPtrArray      *ptr_arrays,
+                     GCopyFunc       element_copy_func,
+                     gpointer        element_copy_data,
+                     GDestroyNotify  element_free_func)
+{
+  g_autoptr(GPtrArray) concatenated = NULL;
+  guint len = 0;
+  guint i = 0;
+  guint j = 0;
+
+  /* O(N) count up of all pointer array lengths */
+  for (i = 0; i < ptr_arrays->len; ++i)
+    {
+      GPtrArray *ptr_array = g_ptr_array_index (ptr_arrays, i);
+      len += ptr_array->len;
+    }
```

Comment:
Hmm. It may be premature optimization on my part but my philosophy here is usually to preallocate when possible and its not too expensive. Even though the allocator probably doubles the allocated region in size every time it needs to grow, repeated reallocation can still be quite expensive for a elements with tens of elements (2, 4, 8, 16, 32, 64)